pkg/apis/workflow/v1alpha1/workflow_phase.go

```diff
@@ -4,17 +4,18 @@ package v1alpha1
 type WorkflowPhase string
 
 const (
-	WorkflowUnknown   WorkflowPhase = ""
-	WorkflowPending   WorkflowPhase = "Pending" // pending some set-up - rarely used
-	WorkflowRunning   WorkflowPhase = "Running" // any node has started; pods might not be running yet, the workflow maybe suspended too
-	WorkflowSucceeded WorkflowPhase = "Succeeded"
-	WorkflowFailed    WorkflowPhase = "Failed" // it maybe that the workflow was terminated
-	WorkflowError     WorkflowPhase = "Error"
+	WorkflowUnknown    WorkflowPhase = ""
+	WorkflowPending    WorkflowPhase = "Pending" // pending some set-up - rarely used
+	WorkflowRunning    WorkflowPhase = "Running" // any node has started; pods might not be running yet, the workflow may be suspended too
+	WorkflowSucceeded  WorkflowPhase = "Succeeded"
+	WorkflowFailed     WorkflowPhase = "Failed" // it may be that the workflow was terminated
+	WorkflowError      WorkflowPhase = "Error"
+	WorkflowCompleting WorkflowPhase = "Completing" // the Workflow completed running but hasn't been fully reconciled yet
```

Comment:
So adding a new `phase` requires changes to _many_ places. That's one reason why we have yet to implement a new `status` like `Cancelled` (see #11849), which has very similar ramifications.

For instance, there are no UI changes in this PR currently, but this does [affect how the UI displays Workflows](https://github.com/argoproj/argo-workflows/blob/66680f1c9bca8b47c40ce918b5d16714058647cb/ui/src/app/shared/utils.ts#L10) -- I don't even know what icon should represent "Completing" as it's not a common UX state.

I did see that there was discussion on whether this should even be user-facing in #12615, as it seems that this state can be derived without a new `phase`

Reply:
thanks for pointing this out @agilgur5

Reply:
If Go had algebraically typed `match` statements, many missed would be picked up by type-checker, but we also have `default`s or `else`s in many places that don't necessarily apply to "Completing" (the UI one is one of those).